format/puffin-spec.md

```diff
@@ -123,6 +123,49 @@ The blob metadata for this blob may include following properties:
 
 - `ndv`: estimate of number of distinct values, derived from the sketch.
 
+#### `delete-vector-v1` blob type
+
+A serialized delete vector (bitmap) that represents the positions of rows in a
+file that are deleted.  A set bit at position P indicates that the row at
+position P is deleted.
+
+The vector supports positive 64-bit positions, but is optimized for cases where
+most positions fit in 32 bits by using a collection of 32-bit Roaring bitmaps.
+64-bit positions are divided into a 32-bit "key" using the most significant 4
+bytes and a 32-bit sub-position using the least significant 4 bytes. For each
+key in the set of positions, a 32-bit Roaring bitmap is maintained to store a
+set of 32-bit sub-positions for that key.
+
+To test whether a certain position is set, its most significant 4 bytes (the
+key) are used to find a 32-bit bitmap and the least significant 4 bytes (the
+sub-position) are tested for inclusion in the bitmap. If a bitmap is not found
+for the key, then it is not set.
+
+The serialized blob contains:
+* The length of the vector and magic bytes stored as 4 bytes, little-endian
```

Comment:
and CRC?

Reply:
No :disappointed:

I also found out that the CRC and this length are big endian. I'm going to look into this more and see what we can do.

Reply:
I love compatibility and all, but that sounds wrong and since we are also including this in the manifests I think we should change it ...

Reply:
I talked with the Delta folks and all implementations rely on on these values to check the validity of the deletion vector. Removing this would mean those checks can't be run and would also break compatibility with all older Delta readers. The length is also intended to make the format self-describing. It is made up of these blobs concatenated together.

The CRC is definitely useful and I don't think there's a reason to make minor changes that break compatibility like swapping the endianness. I'd use the same logic for storing the length. While it isn't necessary when embedding these in Puffin, it isn't such a problem that we should break compatibility with older readers, which would limit options later.

Plus if we keep compatibility, then the deletion vector files in both Delta and Iceberg are the same format with the addition of a Puffin footer and magic bytes. At least it would be an option to use to Puffin as a container format later on.

Reply:
So we will be specifically storing a "delta length" which is the real length - 4 bytes? But in manifests we will keep the full blob length including the crc bytes?

Reply:
> So we will be specifically storing a "delta length" which is the real length - 4 bytes? But in manifests we will keep the full blob length including the crc bytes?

That's... _concerning_ 

> Removing this would mean those checks can't be run and would also break compatibility with all older Delta readers.

I wouldn't be worried about older Delta readers at all.

- compatibility with Delta readers isn't a projec goal, is it?
- compatibility with some old software that we have no control over, shouldn't be a project goal either?

To be clear, it would be nice to reuse valuable concepts from Delta, but we don't need to replicate typos or design issues such as "length is length off by 4".

Reply:
Older Delta readers will never have to read Iceberg Puffin files. Newer software can be made adapted, so that core DV logic is reused even if container formats aren't the same.

Reply:
> So we will be specifically storing a "delta length" which is the real length - 4 bytes? But in manifests we will keep the full blob length including the crc bytes?

I think it would be the blob length - 8 bytes (4 for the CRC and 4 for the length).

> I wouldn't be worried about older Delta readers at all.
> 
> * compatibility with Delta readers isn't a project goal, is it?
> * compatibility with some old software that we have no control over, shouldn't be a project goal either?

These are good questions to consider and we will need to come to consensus as a community.

I think there is value in supporting compatibility with older Delta readers, but I acknowledge that this may be my perspective because my employer has a lot of Delta customers that we are going support now and in the future.

The main use case for maintaining compatibility with the Delta format is that it's hard to move old jobs to new code. We see a similar issue in Hive to Iceberg migrations, too, where unknown older readers prevent migration because they are hard to track down and often read files directly from the backing object store. I'd like to avoid the same problem here, where all readers need to be identified and migrated at the same time. Doing that requires temporarily producing Delta metadata, which this makes possible.

The second reason for maintaining compatibility is that we want for the formats to become more similar. My hope is that we can work across both communities and come up with a common metadata format in a future version. Maintaining compatibility in cases like this keeps our options open: if we have compatible data layers, then it's easier to build a compatible metadata layer.

Other people may not share those goals, but I think the decision comes down to how much is being compromised for compatibility. I don't think it is too much:
* A 4-byte length field that Iceberg doesn't need
* A 4-byte CRC that is useful to add
* Using big endian for this blob (mixed endian if you consider RoaringBitmap is LE)

Reply:
> The main use case for maintaining compatibility with the Delta format is that it's hard to move old jobs to new code. We see a similar issue in Hive to Iceberg migrations, too, where unknown older readers prevent migration because they are hard to track down and often read files directly from the backing object store. I'd like to avoid the same problem here, where all readers need to be identified and migrated at the same time. Doing that requires temporarily producing Delta metadata, which this makes possible.

This will already be the case for an upgrade to Iceberg V3, where as currently proposed all position delete files will be migrated to DV?  Is it technically harder to do a similar migration of Delta lake to a new version if users would want to interop with Iceberg?

Reply:
This is interesting perspective. If indeed we can generate Delta metadata on top of these newly-written Puffin files and feed that to old Delta readers without any modifications to those old readers, there's certainly some value in it.

@rdblue this discussion is interesting and compatibility capabilities are _potentially_ useul.
Why isn't there "Delta" word in the spec and only one ([unclear](https://github.com/apache/iceberg/pull/11238/files#r1796159961)) occurrence of "compatibility"? wouldn't it be useful for better understanding of a) why it's the way it is and b) what a user can do thanks to this?
Without this the spec is complete (context isn't needed for completeness), but confusing. There are some redundant fields (eg length) and a spec implementor may spend some time wondering eg  whether length needs to be validated or ignored.